examples/mxnet_imagenet_resnet50.py

```diff
@@ -0,0 +1,290 @@
+        return self
+    @property
+    def provide_data(self):
+        return [mx.io.DataDesc('data', self.data.shape, self.dtype)]
+    @property
+    def provide_label(self):
+        return [mx.io.DataDesc('softmax_label', (self.batch_size,), self.dtype)]
+    def next(self):
+        self.cur_iter += 1
+        if self.cur_iter <= self.max_iter:
+            return DataBatch(data=(self.data,),
+                             label=(self.label,),
+                             pad=0,
+                             index=None,
+                             provide_data=self.provide_data,
+                             provide_label=self.provide_label)
+        else:
+            raise StopIteration
+    def __next__(self):
+        return self.next()
+    def reset(self):
+        self.cur_iter = 0
+
+# Logs eval metrics at the end of an epoch
+class LogValidationMetricsCallback(object):
+
+    def __call__(self, param):
+        if not param.eval_metric:
+            return
+        name_value = param.eval_metric.get_name_value()
+        for name, value in name_value:
+            val_array = mx.nd.array([value])
+            hvd.allreduce_(val_array, average=True, name=str(-1))
+            logging.info('Epoch[%d] Validation-%s=%f', param.epoch, name, val_array.asscalar())
+
+# Fetch training and validation data if present
+try:
+    train_data, val_data, batch_fn = get_data_rec(args.rec_train, args.rec_train_idx,
+                                              args.rec_val, args.rec_val_idx,
+                                              batch_size, data_nthreads)
+except Exception as e:
+    print("Data not found, using synthetic data instead!")
+    args.benchmark = 1
+
+# Otherwise use synthetic data
+if args.benchmark is 1:
+    image_shape = (3, 224, 224)
+    data_shape = (batch_size,) + image_shape
+    train_data  = SyntheticDataIter(classes, data_shape, epoch_size, np.float32)
+    val_data = None
+
+def main():
+    # Get model from Gluon model zoo
+    net = get_model(args.model, **kwargs)
+    net.cast(args.dtype)
+
+    # Create input symbol
+    data = mx.sym.var('data')
+    if args.dtype == 'float16':
+        data = mx.sym.Cast(data=data, dtype=np.float16)
+        net.cast(np.float16)
+
+    # Create output symbol
+    out = net(data)
+    if args.dtype == 'float16':
+        out = mx.sym.Cast(data=out, dtype=np.float32)
+    softmax = mx.sym.SoftmaxOutput(out, name='softmax')
+
+    if args.use_pretrained:
+        arg_params = {} 
+        for x in net.collect_params().values():
+            x.reset_ctx(mx.cpu())
+            arg_params[x.name] = x.data()
+    else:
+        arg_params = None
+    aux_params = None
+
+    # Create model
+    mod = mx.mod.Module(softmax, context=context)
+
+    # Create optimizer
+    optimizer_params = {'wd': args.wd,
+                        'momentum': args.momentum,
+                        'lr_scheduler': lr_scheduler,
+                        'rescale_grad': 1.0/batch_size/num_workers,
```

Comment:
The reason is that we have seen some performance degradation with doing gradient averaging in the DistributedOptimizer. It seems to happen, because of a `wait_to_read()` on the NDArray [here](https://github.com/uber/horovod/blob/645cbd0125b3556da531f8d5c1989a1308fb5c4b/horovod/mxnet/mpi_ops.py#L86). This `wait_to_read()` is necessary, because we noticed that the elementwise operations (that follows it on Line 87) get captured in the task graph generated by the scheduler. The two enter a race condition, which generates wrong results half the time.

More details in this commit: [0517fa0](https://github.com/uber/horovod/pull/542/commits/0517fa0289b0fdc12e3b7237b5ec4e5b39e98a1b)

Reply:
It's not very user-friendly to push it down on user though.  How large is the degradation?

I'm still reading the code, but could this averaging be done in the callback, like it's done for PyTorch?